File: mobile/src/test/java/org/openhab/habdroid/core/connection/DefaultConnectionTest.java
package org.openhab.habdroid.core.connection;

import okhttp3.Credentials;
import okhttp3.OkHttpClient;
import org.junit.Before;
import org.junit.Test;
import org.openhab.habdroid.util.AsyncHttpClient;
import org.openhab.habdroid.util.HttpClient;
import org.openhab.habdroid.util.SyncHttpClient;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;

public class DefaultConnectionTest {
    private static final String TEST_BASE_URL = "https://demo.local:8443";

    private OkHttpClient mClient;

    private Connection mTestConnection;
    private Connection mTestConnectionRemote;
    private Connection mTestConnectionCloud;

    @Before
    public void setup() {
        mClient = new OkHttpClient.Builder().build();
        mTestConnection = new DefaultConnection(mClient, Connection.TYPE_LOCAL,
                TEST_BASE_URL, null, null);
        mTestConnectionRemote = new DefaultConnection(mClient, Connection.TYPE_REMOTE,
                null, null, null);
        mTestConnectionCloud = new DefaultConnection(mClient,
                Connection.TYPE_CLOUD, null, null, null);
    }

    @Test
    public void testGetConnectionTypeSetRemote() {
        assertEquals(Connection.TYPE_REMOTE, mTestConnectionRemote.getConnectionType());
    }

    @Test
    public void testGetConnectionTypeSetLocal() {
        assertEquals(Connection.TYPE_LOCAL, mTestConnection.getConnectionType());
    }

    @Test
    public void testGetConnectionTypeSetCloud() {
        assertEquals(Connection.TYPE_CLOUD, mTestConnectionCloud.getConnectionType());
    }

    @Test
    public void testGetUsernameNotSet() {
        assertNull(mTestConnection.getUsername());
    }

    @Test
    public void testGetPasswordNotSet() {
        assertNull(mTestConnection.getPassword());
    }

    @Test
    public void testGetUsernameSet() {
        Connection connection = new DefaultConnection(mClient, Connection.TYPE_LOCAL,
                TEST_BASE_URL, "Test-User", null);
        assertEquals("Test-User", connection.getUsername());
    }

    @Test
    public void testGetPasswordSet() {
        Connection connection = new DefaultConnection(mClient, Connection.TYPE_LOCAL,
                TEST_BASE_URL, null, "Test-Password");
        assertEquals("Test-Password", connection.getPassword());
    }

    @Test
    public void testGetSyncHttpClientCached() {
        SyncHttpClient client1 = mTestConnection.getSyncHttpClient();
        SyncHttpClient client2 = mTestConnection.getSyncHttpClient();

        assertNotNull(client1);
        assertEquals(client1, client2);
    }

    @Test
    public void testGetAsyncHttpClientCached() {
        AsyncHttpClient client1 = mTestConnection.getAsyncHttpClient();
        AsyncHttpClient client2 = mTestConnection.getAsyncHttpClient();

        assertNotNull(client1);
        assertEquals(client1, client2);
    }

    @Test
    public void testAsyncHasNoUsernamePassword() {
        HttpClient httpClient = mTestConnection.getAsyncHttpClient();
        assertNull(httpClient.mAuthHeader);
    }

    @Test
    public void testSyncHasNoUsernamePassword() {
        HttpClient httpClient = mTestConnection.getSyncHttpClient();
        assertNull(httpClient.mAuthHeader);
    }

    @Test
    public void testAsyncHasUsernamePassword() {
        Connection connection = new DefaultConnection(mClient, Connection.TYPE_LOCAL,
                TEST_BASE_URL, "Test-User", "Test-Password");
        HttpClient httpClient = connection.getAsyncHttpClient();

        assertEquals(Credentials.basic("Test-User", "Test-Password"),
                httpClient.mAuthHeader);
    }

    @Test
    public void testSyncHasUsernamePassword() {
        Connection connection = new DefaultConnection(mClient, Connection.TYPE_LOCAL,
                TEST_BASE_URL, "Test-User", "Test-Password");
        HttpClient httpClient = connection.getSyncHttpClient();

        assertEquals(Credentials.basic("Test-User", "Test-Password"),
                httpClient.mAuthHeader);
    }

    @Test
    public void testSyncResolveRelativeUrl() {
        SyncHttpClient.HttpResult result = mTestConnection.getSyncHttpClient().get("rest/test");
        assertFalse("The request should never succeed in tests", result.isSuccessful());
        assertEquals(TEST_BASE_URL + "/rest/test", result.request.url().toString());
        result.close();

        result = mTestConnection.getSyncHttpClient().get("/rest/test");
        assertEquals(TEST_BASE_URL + "/rest/test", result.request.url().toString());
        result.close();
    }

    @Test
    public void testSyncResolveAbsoluteUrl() {
        SyncHttpClient.HttpResult result =
                mTestConnection.getSyncHttpClient().get("http://mylocalmachine.local/rest/test");
        assertFalse("The request should never succeed in tests", result.isSuccessful());
        assertEquals("http://mylocalmachine.local/rest/test", result.request.url().toString());
        result.close();
    }
}

File: mobile/src/test/java/org/openhab/habdroid/core/connection/DefaultConnectionTest.kt
/*
 * Copyright (c) 2010-2020 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.core.connection

import kotlinx.coroutines.runBlocking
import okhttp3.Credentials
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.openhab.habdroid.model.ServerPath
import org.openhab.habdroid.util.HttpClient

class DefaultConnectionTest {
    private lateinit var client: OkHttpClient

    private lateinit var testConnection: Connection
    private lateinit var testConnectionRemote: Connection
    private lateinit var testConnectionCloud: Connection

    @Before
    fun setup() {
        client = OkHttpClient.Builder().build()
        testConnection = DefaultConnection(
            client,
            Connection.TYPE_LOCAL,
            ServerPath(TEST_BASE_URL, null, null)
        )
        testConnectionRemote = DefaultConnection(
            client,
            Connection.TYPE_REMOTE,
            ServerPath("", null, null)
        )
        testConnectionCloud = DefaultConnection(
            client,
            Connection.TYPE_CLOUD,
            ServerPath("", null, null)
        )
    }

    @Test
    fun testGetConnectionTypeSetRemote() {
        assertEquals(Connection.TYPE_REMOTE, testConnectionRemote.connectionType)
    }

    @Test
    fun testGetConnectionTypeSetLocal() {
        assertEquals(Connection.TYPE_LOCAL, testConnection.connectionType)
    }

    @Test
    fun testGetConnectionTypeSetCloud() {
        assertEquals(Connection.TYPE_CLOUD, testConnectionCloud.connectionType)
    }

    @Test
    fun testGetUsernameNotSet() {
        assertNull(testConnection.username)
    }

    @Test
    fun testGetPasswordNotSet() {
        assertNull(testConnection.password)
    }

    @Test
    fun testGetUsernameSet() {
        val connection = DefaultConnection(
            client,
            Connection.TYPE_LOCAL,
            ServerPath(TEST_BASE_URL, "Test-User", null)
        )
        assertEquals("Test-User", connection.username)
    }

    @Test
    fun testGetPasswordSet() {
        val connection = DefaultConnection(
            client,
            Connection.TYPE_LOCAL,
            ServerPath(TEST_BASE_URL, null, "Test-Password")
        )
        assertEquals("Test-Password", connection.password)
    }

    @Test
    fun testGetHttpClientCached() {
        val client1 = testConnection.httpClient
        val client2 = testConnection.httpClient

        assertNotNull(client1)
        assertEquals(client1, client2)
    }

    @Test
    fun testHasNoUsernamePassword() {
        val httpClient = testConnection.httpClient
        assertNull(httpClient.authHeader)
    }

    @Test
    fun testHasUsernamePassword() {
        val connection = DefaultConnection(
            client,
            Connection.TYPE_LOCAL,
            ServerPath(TEST_BASE_URL, "Test-User", "Test-Password")
        )
        val httpClient = connection.httpClient

        assertEquals(Credentials.basic("Test-User", "Test-Password"), httpClient.authHeader)
    }

    @Test
    fun testResolveRelativeUrl() {
        val requestUrl1 = getRequestUrlForUrl("rest/test")
        assertEquals("$TEST_BASE_URL/rest/test", requestUrl1.toString())

        val requestUrl2 = getRequestUrlForUrl("/rest/test")
        assertEquals("$TEST_BASE_URL/rest/test", requestUrl2.toString())
    }

    @Test
    fun testSyncResolveAbsoluteUrl() {
        val requestUrl = getRequestUrlForUrl("http://mylocalmachine.local/rest/test")
        assertEquals("http://mylocalmachine.local/rest/test", requestUrl.toString())
    }

    private fun getRequestUrlForUrl(url: String): HttpUrl = try {
        runBlocking {
            val result = testConnection.httpClient.get(url)
            result.close()
            assertFalse("The request should never succeed in tests", true)
            result.request.url
        }
    } catch (e: HttpClient.HttpException) {
        e.request.url
    }

    companion object {
        private const val TEST_BASE_URL = "https://demo.local:8443"
    }
}
